libraries/lib-wave-track/WaveTrack.cpp

```diff
@@ -2339,6 +2339,202 @@ bool WaveTrack::Get(size_t iChannel, size_t nBuffers,
    });
 }
 
+namespace {
+void RoundToNearestClipSample(const WaveTrack& track, double& t)
+{
+   const auto clip = track.GetClipAtTime(t);
+   if (!clip)
+      return;
+   t = clip->SamplesToTime(clip->TimeToSamples(t - clip->GetPlayStartTime())) +
+       clip->GetPlayStartTime();
+}
+}
+
+std::pair<size_t, size_t> WaveTrack::GetFloatsCenteredAroundTime(
+   double t, size_t iChannel, float* buffer, size_t numSideSamples,
+   bool mayThrow) const
+{
+   const auto numSamplesReadLeft = GetFloatsFromTime(
+      t, iChannel, buffer, numSideSamples, mayThrow, PlaybackDirection::backward);
+   const auto numSamplesReadRight = GetFloatsFromTime(
+      t, iChannel, buffer + numSideSamples, numSideSamples + 1, mayThrow,
+      PlaybackDirection::forward);
+   return { numSideSamples - numSamplesReadLeft,
+            numSideSamples + numSamplesReadRight };
+}
+
+namespace
+{
+template<typename BufferType>
+struct BufferCharType;
+
+template<>
+struct BufferCharType<float*>
+{
+   using type = samplePtr;
+};
+
+template<>
+struct BufferCharType<const float*>
+{
+   using type = constSamplePtr;
+};
+
+template <typename BufferType> struct SampleAccessArgs
+{
+   const typename BufferCharType<BufferType>::type offsetBuffer;
+   const sampleCount start;
+   const size_t len;
+};
+
+template <typename BufferType>
+SampleAccessArgs<BufferType> GetSampleAccessArgs(
+   const WaveClip& clip, double startOrEndTime /*absolute*/, BufferType buffer,
+   size_t totalToRead, size_t alreadyRead, bool forward)
+{
+   const auto remainingToRead = totalToRead - alreadyRead;
+   const auto sampsInClip = clip.GetVisibleSampleCount();
+   const auto sampsPerSec = clip.GetRate() / clip.GetStretchRatio();
+   if (forward)
+   {
+      const auto startTime =
+         std::max(startOrEndTime - clip.GetPlayStartTime(), 0.);
+      const sampleCount startSamp { std::ceil(
+         std::round(startTime * sampsPerSec * 1e6) / 1e6) };
+      if (startSamp >= sampsInClip)
+         return { nullptr, sampleCount { 0u }, 0u };
+      const auto len =
+         limitSampleBufferSize(remainingToRead, sampsInClip - startSamp);
+      return { reinterpret_cast<typename BufferCharType<BufferType>::type>(
+                  buffer + alreadyRead),
```

Comment:
In the case that `max` above clipped a negative value to 0,  should the buffer instead get some zeroes at `buffer + alreadyRead` (if writing), and should the first member of the return be a later position?

Maybe not but I'm trying to understand the usage

Reply:
I'm not sure I understand your question.
`alreadyRead` means `[buffer, buffer + alreadyRead)` are indices that already have been dealt with.
`startSamp` tells which first sample index of the clip matches `buffer + alreadyRead`.